logstash-core/src/main/java/org/logstash/config/ir/CompiledPipeline.java

```diff
@@ -312,30 +317,94 @@ private boolean isOutput(final Vertex vertex) {
          */
         private final Map<String, Dataset> plugins = new HashMap<>(50);
 
-        private final Dataset compiled;
+        private final Dataset compiledFilters;
+        private final Dataset compiledOutputs;
 
         CompiledExecution() {
-            compiled = compile();
+            compiledFilters = compileFilters();
+            compiledOutputs = compileOutputs();
+        }
+
+        @SuppressWarnings("rawtypes")
+        public void compute(final RubyArray batch, final boolean flush, final boolean shutdown) {
+            unorderedCompute(batch, flush, shutdown);
         }
 
-        Dataset toDataset() {
-            return compiled;
+        @SuppressWarnings("rawtypes")
+        public void unorderedCompute(final RubyArray batch, final boolean flush, final boolean shutdown) {
+            @SuppressWarnings("unchecked") final RubyArray<JrubyEventExtLibrary.RubyEvent> outputBatch = RubyUtil.RUBY.newArray();
+            copyNonCancelledEvents(compiledFilters.compute(batch, flush, shutdown), outputBatch);
```

Comment:
nitpick: I would rather see a local variable here with the result of the computation, which is then passed to `copyNonCancelledEvents`. There is a lot going on, and especially because some of these functions modify the passed collections in-place, it is easy to lose track of the objects.

Reply:
exposed result.